File: Cpp/206.cpp
#include <iostream>
#include <vector>

#include "ListNode.h"

using namespace std;

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* temp = head;

        while (head) {
            temp = head->next;
            head->next = prev;
            prev = head;
            head = temp;
        }

        return prev;
    }
};

class Solution2 {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }

        ListNode* tail = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return tail;
    }
};

int main() {
    vector<int> l = {1, 2, 3, 4, 5};
    ListNodeEnv env;
    ListNode* head = env.initialize(l);
    cout << "before reverse the linked list:" << endl;
    env.showList(head);

    Solution2 s;
    ListNode* new_head = s.reverseList(head);
    cout << "after reverse the linked list:" << endl;
    env.showList(new_head);

    return 0;
}

File: Python/206.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        prev = None
        while head != None:
            next = head.next
            head.next = prev
            prev = head
            head = next
        return prev



class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        return self.reverse(head)
        
    def reverse(self, head, prev=None):
        if head == None:
            return prev
        next = head.next
        head.next = prev
        return self.reverse(next, head)
